File: CP/LeetCode/Python/38.py
# 38. Count and Say

from __future__ import annotations


class Solution:
    # def count(self, s: str) -> str:
    #     res = ""
    #     count = 1
    #     for i, number in enumerate(s):
    #         if i < len(s) - 1 and number == s[i + 1]:
    #             count += 1
    #         else:
    #             res += str(count) + number
    #             count = 1
    #     return res

    def countAndSay(self, n: int) -> str:
        # if n == 1:
        #     return "1"

        # return self.count(self.countAndSay(n - 1))

        if n == 1:
            return "1"

        prev = self.countAndSay(n - 1)
        res = ""
        count = 1
        for i, number in enumerate(prev):
            if i < len(prev) - 1 and number == prev[i + 1]:
                count += 1
            else:
                res += str(count) + number
                count = 1
        return res


if __name__ == "__main__":
    solution = Solution()
    number = 4
    print(solution.countAndSay(number))

File: CP/LeetCode/Java/38.java
// 38. Count and Say

class Solution {
    public String countAndSay(final int n) {
        if (n == 1) {
            return "1";
        }
        String prev = countAndSay(n - 1);
        StringBuilder sb = new StringBuilder();
        int count = 1;
        for (int i = 1; i < prev.length(); i++) {
            if (prev.charAt(i) == prev.charAt(i - 1)) {
                count++;
            } else {
                sb.append(count);
                sb.append(prev.charAt(i - 1));
                count = 1;
            }
        }
        sb.append(count);
        sb.append(prev.charAt(prev.length() - 1));
        return sb.toString();
    }

    public static void main(final String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.countAndSay(4));
    }
}
